pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+// Trie is a non-thread-safe binary trie that indexes arbitrarily long
+// bit-based keys with associated prefixes (that is, "masks") indexed from
+// [most significant bit] ("MSB") to [least significant bit] ("LSB") using
+// the [longest prefix match algorithm].
+//
+// Each method's comments describes the mechanism of how the method
+// works.
+//
+// [most significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Most_significant_bit
+// [least significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit
+// [longest prefix match algorithm]: https://en.wikipedia.org/wiki/Longest_prefix_match
+type Trie[K, T any] interface {
+	// Update updates the trie with a a prefix, key, and value.
+	//
+	// Note:If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Upsert(prefix uint, key K, value T)
```

Comment:
`uint` is 32 bits on 32 bit platforms, and 64 bits on 64 bit platforms. Should this be `uint64`?

Also, what about ipv6 :-(

Reply:
(unless this is the bit-length and not the bit prefix of the field, in which case, nevermind)

Reply:
Maybe update the function comment to describe prefix in more detail, or provide an example of its usage?

Reply:
Fair enough. `prefix` is the mask length to add to the key. It is unit so that arbitrarily long prefixes can be added. The key is where the data is stored and can undoubtedly be longer than uint64 (I have a big-endian key implementation that I haven't added to this PR).